[XLA:GPU][ROCm] Enable XLA JIT compiler on ROCm.

whchung opened the pull request:
- bazel changes to enable xla_gpu_device and xla_gpu_jit on ROCm.
- Disable cusolver_context on ROCm. It has source code dependency to CUDA API.
- Disable dependency to cholesky_thunk on ROCm. It has source code dependency to CUDA API.
- Remove cudnn_conv_algorithm_picker from gpu_compiler_impl dependency list. It is conditionally dependent when CUDA is enabled.
- Remove CUDA-specific header inclusions in collective_permute_thunk and custom_call_thunk. These 2 thunks actually work on ROCm.
- Partially enable ptxas_utils.h to make things build on ROCm. Full-fledged solution is on PR #30884.

whchung committed:
Enable XLA JIT compiler on ROCm.

- bazel changes to enable xla_gpu_device and xla_gpu_jit on ROCm.
- Disable cusolver_context on ROCm. It has source code dependency to CUDA API.
- Disable dependency to cholesky_thunk on ROCm. It has source code dependency
  to CUDA API.
- Remove cudnn_conv_algorithm_picker from gpu_compiler_impl dependency list.
  It is conditionally dependent when CUDA is enabled.
- Remove CUDA-specific header inclusions in collective_permute_thunk and
  custom_call_thunk. These 2 thunks actually work on ROCm.
- Partially enable ptxas_utils.h to make things build on ROCm. Full-fledged
  solution is on PR #30884.

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -464,6 +466,7 @@ Status IrEmitterUnnested::HandleCustomCall(HloInstruction* custom_call) {
     return Status::OK();
   }
 
+#if !TENSORFLOW_USE_ROCM
```
It's unfortunate that the XLA compiler has a CUDA header dependency. It shouldn't be there. Anyway, using `#if` is justified in this case. Can you use something like `#if TENSORFLOW_USE_CUDA` rather than `!TENSORFLOW_USE_ROCM`? That'd be cleaner in my opinion?

whchung replied:
@thomasjoerg need your guidance here. there is really no such macro as `TENSORFLOW_USE_CUDA` but there is `GOOGLE_CUDA`.

but in our earlier PRs it seems use of `GOOGLE_CUDA` is disallowed within XLA so I resorted to use `!TENSORFLOW_USE_ROCM`.

whchung replied:
a quick grep within XLA shows only `nccl_all_reduce_thunk.cc` has references to `GOOGLE_CUDA` macro.

mellanox-github commented:
Can one of the admins verify this patch?

whchung committed:
Skip including cholesky_thunk in thunk_emitter on ROCm build.

cholesky_thunk has source code dependency to cusolver_context which has source
code dependency to CUDA-specific headers. Disable them on ROCm build.

Since GOOGLE_CUDA macros is discouraged in XLA, use !TENSORFLOW_USE_ROCM
instead.

cheshire commented:
@whchung @thomasjoerg Is on a leave for the next couple of weeks. I can help to get this patch merged. Could you rebase and resolve the conflict?

whchung commented:
@cheshire thanks for the heads up let me rebase real quick.

cheshire commented:
@whchung I've had to change quite a few things, please take a look at the patch which has landed.
